### PR TITLE
NAS-108926 / 21.XX / Enhanced upgrade mechanism for Enterprise

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

**Review by william-gr — COMMENTED** [review]
We will also need an alert if we are running a non enterprise ready version on enterprise hardware

**Comment by themylogin** [comment]
@william-gr so it will issue an alert for those N% systems who were updated first?

**Comment by william-gr** [comment]
@themylogin No. Like said in the ticket, a build should be able to be flagged as enterprise ready or not, which can be overridden via update server.

Running a version that was built without that flag or doesnt have override, issue an alert.

**Comment by themylogin** [comment]
@william-gr so it should a matter of presence of e.g. `/usr/local/share/truenas/enterprise-ready` file? And enterprise-ready and non-enterprise-ready images will be different from each other?

**Comment by william-gr** [comment]
@themylogin Either that or a new attribute in the /data/manifest

No, we wont have different images, either an image is enterprise ready or not (Which could be overridden via update server)

**Comment by themylogin** [comment]
@william-gr so that alert will poll update server periodically to know if current version is considered enterprise ready?

**Comment by william-gr** [comment]
@themylogin I was thinking not periodically, but whenever we check if there is a new update available

**Review by william-gr — COMMENTED** [review]
What if this is an air-gapped install. How do I know if this is enterprise ready version or not?

**Comment by themylogin** [comment]
@william-gr if current version is greater than latest enterprise-ready, then it is not enterprise ready (and I suppose air-gapped install is the only way to obtain such a version)

**Comment by william-gr** [comment]
@themylogin I dont follow, maybe I didnt phrase this correctly.

1. Install TrueNAS 12.0-U2 -- that build is not enterprise ready
2. I have no internet connection
3. How do I get the alert the version I am running is not enterprise ready?

**Comment by themylogin** [comment]
@william-gr you won't, because you've said we want the same release to be not enterprise-ready and then became enterprise-ready (so we won't be able to change any flag files in the filesystem, so we'll have to rely on the internet).

**Comment by william-gr** [comment]
@themylogin No, I said we want it to be able to identify via a flag in the build. I said we don't want two different builds for the same release (one for CORE and one for ENTERPRISE)

**Comment by themylogin** [comment]
@william-gr so we release e.g. an U4 without enterprise-ready flag, then we fix it and add enterprise-ready flag and just replace U4 file?

**Comment by william-gr** [comment]
No, either a version is enterprise ready or not, no replacing anything

**Comment by themylogin** [comment]
@william-gr so we release not enterprise-ready U4, then fix it and release a U5 that is enterprise-ready?

**Comment by william-gr** [comment]
Yes. For air-gapped installs anyway. For online update we could potentially mark an image as enterprise ready, overriding the build flag

**Comment by themylogin** [comment]
@william-gr so we continue using current method but if network fails then we rely on a filesystem flag?

**Comment by william-gr** [comment]
@themylogin Yup

**Review by william-gr — APPROVED** [review]

**Comment by rick-mesta** [comment]
@william-gr / @themylogin: Looks like you fellas resolved this issue and this is ready to be merged, correct ? I'd appreciate a 👍🏼 or 👎🏼 from both -- just to make sure I don't have to revert stuffs 😝

**Comment by william-gr** [comment]
@rick-mesta I believe so, please notify Joe when this is merged so he is aware and can plan testing for this new feature. It will need some extra planning.
It may also need core-build repo merge, was it done already @themylogin ?

**Comment by rick-mesta** [comment]
@william-gr: will definitely let @pkgdemon know once this delivers... still waiting on @themylogin's go/no-go before I make any moves 😏

**Comment by themylogin** [comment]
@william-gr there is a corresponding PR here https://github.com/truenas/freenas-pkgtools/pull/28

And building/publishing images was out of my scope IIRC

File: src/middlewared/middlewared/plugins/update_/download_freebsd.py [hunk 1]
```diff
@@ -32,8 +32,10 @@ def download_impl(self, job, train, location, progress_proportion):
             location,
             check_handler=handler.check_handler,
             get_handler=handler.get_handler,
+            **self.middleware.call_sync('update.enterprise_kwargs'),
```
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/plugins/update_/download_freebsd.py]: Good catch, thank you

**Review by yocalebo — APPROVED** [review]

**Comment by pkgdemon** [comment]
@rick-mesta Remind me next week we can make a special custom build using the PR's (we don't to merge to test).  I will plan to have systems free by Wednesday to deploy it on, and Tim Moore should have a testing plan with first thing next week that we can review ahead of this as well.

**Pull request closed without merging**